pkg/statuschecker/types/built_in.go

```diff
@@ -66,23 +67,65 @@ func alwaysReady(_ runtime.Object) (statuschecker.ObjectStatusResult, error) {
 	return statuschecker.ObjectStatusReady{}, nil
 }
 
+func getDeploymentCondition(deployment apps_v1.Deployment, condType apps_v1.DeploymentConditionType) *apps_v1.DeploymentCondition {
+	deploymentStatus := deployment.Status
+	for i := range deploymentStatus.Conditions {
+		c := deploymentStatus.Conditions[i]
+		if c.Type == condType {
+			return &c
+		}
+	}
+	return nil
+}
+
 // Works according to https://kubernetes.io/docs/user-guide/deployments/#the-status-of-a-deployment
-// and k8s.io/kubernetes/pkg/client/unversioned/conditions.go:120 DeploymentHasDesiredReplicas()
 func isDeploymentReady(obj runtime.Object) (statuschecker.ObjectStatusResult, error) {
+
 	var deployment apps_v1.Deployment
 	if err := util.ConvertType(appsV1Scheme, obj, &deployment); err != nil {
 		return nil, err
 	}
 
-	replicas := int32(1) // Default value if not specified
-	if deployment.Spec.Replicas != nil {
-		replicas = *deployment.Spec.Replicas
-	}
+	replicas := deployment.Spec.Replicas
+
+	generation := deployment.Generation
+	observedGeneration := deployment.Status.ObservedGeneration
+	updatedReplicas := deployment.Status.UpdatedReplicas
+	availableReplicas := deployment.Status.AvailableReplicas
+
+	if generation <= observedGeneration {
+		progressingCond := getDeploymentCondition(deployment, apps_v1.DeploymentProgressing)
+		if progressingCond != nil && progressingCond.Reason == timedOutReason {
+			return statuschecker.ObjectStatusError{
+				UserError:      false,
```

Comment:
Could we flip this to "true"? I might rename `UserError` so it's less badly named.

Effectively this is still a "user error" as the deployment failed because the user specified a bad image, or the healthchecks failed, or they requested too many pods, etc.

Reply:
(it's really more like "NonSmithError" - it's not an error that resulted because something unexpected happened in smith)

Reply:
Done